File: solutions/cpp/binary-tree-postorder-traversal-iterative.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <memory>
#include <stack>
#include <vector>

#include "./Binary_tree_prototype.h"
#include "./Binary_tree_utils.h"

using std::cout;
using std::endl;
using std::equal;
using std::stack;
using std::unique_ptr;
using std::vector;

// @include
// We use stack and previous node pointer to simulate postorder traversal.
vector<int> PostorderTraversal(const unique_ptr<BinaryTreeNode<int>>& root) {
  if (!root) {  // Empty tree.
    return {};
  }

  stack<BinaryTreeNode<int>*> s;
  BinaryTreeNode<int>* prev = nullptr;
  s.emplace(root.get());
  vector<int> res;
  while (!s.empty()) {
    auto curr = s.top();
    if (!prev || prev->left.get() == curr || prev->right.get() == curr) {
      // Going down.
      if (curr->left) {  // Visit left.
        s.emplace(curr->left.get());
      } else if (curr->right) {  // Visit right.
        s.emplace(curr->right.get());
      } else {  // Leaf node, then process current node.
        res.emplace_back(curr->data);
        s.pop();
      }
    } else if (curr->left.get() == prev) {
      // Going up, finished visiting left.
      if (curr->right) {  // Visit right.
        s.emplace(curr->right.get());
      } else {  // No right child, then process current node.
        res.emplace_back(curr->data);
        s.pop();
      }
    } else {  // curr->right.get() == prev.
      // Going up, finished visiting left and right.
      res.emplace_back(curr->data);
      s.pop();
    }
    prev = curr;
  }
  return res;
}
// @exclude

int main(int argc, char** argv) {
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> root = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{3, nullptr, nullptr});
  root->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{2, nullptr, nullptr});
  root->left->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{1, nullptr, nullptr});
  root->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{5, nullptr, nullptr});
  root->right->left = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{4, nullptr, nullptr});
  root->right->right = unique_ptr<BinaryTreeNode<int>>(
      new BinaryTreeNode<int>{6, nullptr, nullptr});
  auto res = PostorderTraversal(root), golden_res = generate_postorder(root);
  assert(res.size() == golden_res.size() &&
         equal(res.cbegin(), res.cend(), golden_res.cbegin()));
  return 0;
}

File: solutions/python/binary_tree_postorder_traversal_iterative.py
from binary_tree_prototype import BinaryTreeNode
from binary_tree_utils import generate_postorder


# @include
# We use stack and previous node pointer to simulate postorder traversal.
def postorder_traversal(tree):
    if not tree:  # Empty tree.
        return []

    path_stack, prev, postorder_sequence = [tree], None, []
    while path_stack:
        curr = path_stack[-1]
        if not prev or prev.left is curr or prev.right is curr:
            # We came down to curr from prev.
            if curr.left:  # Traverse left.
                path_stack.append(curr.left)
            elif curr.right:  # Traverse right.
                path_stack.append(curr.right)
            else:  # Leaf node, visit current node.
                postorder_sequence.append(curr.data)
                path_stack.pop()
        elif curr.left is prev:
            # Done with left, now traverse right.
            if curr.right:
                path_stack.append(curr.right)
            else:  # No right child, so visit curr.
                postorder_sequence.append(curr.data)
                path_stack.pop()
        else:
            # Finished traversing left and right, so visit curr.
            postorder_sequence.append(curr.data)
            path_stack.pop()
        prev = curr
    return postorder_sequence
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode(3)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(1)
    tree.right = BinaryTreeNode(5)
    tree.right.left = BinaryTreeNode(4)
    tree.right.right = BinaryTreeNode(6)
    res = postorder_traversal(tree)
    golden_res = generate_postorder(tree)
    assert res == golden_res


if __name__ == '__main__':
    main()
